Update README (#6)
Prolegis/congress: README.md

# Prolegis/congress

> This is a fork of the UnitedStatesIO Congress repository (unitedstates/congress). It is living on an EC2 instance [here](https://us-east-1.console.aws.amazon.com/ec2/home?region=us-east-1#InstanceDetails:instanceId=i-0e99e6fa752b897ab).

Reviewer: Suggestion to link directly to the repo we're forking
```suggestion
This repo is a fork of [unitedstates/congress](https://github.com/unitedstates/congress), hosted by Prolegis on an EC2 instance [here](https://us-east-1.console.aws.amazon.com/ec2/home?region=us-east-1#InstanceDetails:instanceId=i-0e99e6fa752b897ab).
```

I'm also not sure it's a good idea to link directly to our EC2, but I could just be paranoid

Author: Addressed here: https://github.com/Prolegis/congress/pull/6/commits/6417f8e9ce2a3859f708061668492f35c3b0c87a#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R3

I removed the link to the EC2 instance.

This repository was originally developed by [GovTrack.us](https://www.govtrack.us) and the Sunlight Foundation in 2013 (see [Eric's blog post](https://sunlightfoundation.com/blog/2013/08/20/a-modern-approach-to-open-data/)) and is currently maintained by GovTrack.us and other contributors. For more information about data in Congress, see the [Congressional Data Coalition](https://congressionaldata.org/).

## EC2 Configuration
The purpose of this repository is to provide tools for generating congressional data that can be integrated into our Rails application. Since these tools are designed for local machine use, we cannot simply host the repository as an API to make requests.

> Instead, we host the entire repository on S3, enabling the following functionalities:

Reviewer: > we host the entire repository on S3

Is this right?

Author: Addressed here: https://github.com/Prolegis/congress/commit/6417f8e9ce2a3859f708061668492f35c3b0c87a#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R24

* Generating congressional data using the tools in this repository.
* Uploading the generated data to S3.
* Triggering requests to import this data into the Rails application (e.g., having the Rails app asynchronously read from S3 after hitting a specific endpoint).

### AWS Configuration
To enable S3 uploads, configure AWS access by running the following command in the terminal: `aws configure`

You will be prompted to enter:
**1** **AWS Access Key ID**
**2** **AWS Secret Access Key**

> ⠀These credentials can be found in ~[1Password](https://my.1password.com/app#/everything/Search/p7c45mgv6bof7c5pagpgc3nolacebxppcpe2r47tl3df5lh4s4ja?itemListId=Congress+Repo)~.

Reviewer: Just me being paranoid, but I don't think we should ever link directly to anything in 1pass, it should be sufficient to just tell people where to go looking around

Author: Addressed here: https://github.com/Prolegis/congress/pull/6/commits/6417f8e9ce2a3859f708061668492f35c3b0c87a#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L36

### Environment Variables
Set the following environment variables in the ~/.bashrc file to ensure proper functionality:
* API_KEY_PRODUCTION - Enables interaction with the production Rails application.
* API_KEY_STAGING - Enables interaction with the staging Rails application.
* API_KEY_DEMO - Enables interaction with the demo Rails application.
* GITHUB_PERSONAL_ACCESS_TOKEN - Enables cloning the latest version of the repository.

> The GitHub token will expire a year from when this is written (Jan 7, 2025). In the next year, it will need to be replaced. This token grants this EC2 instance only the ability to pull down the Prolegis/congress repo. Eventually, we should add a GitHub action so that when Prolegis/congress is updated, the changes get deployed to this EC2 server. 

Reviewer: ```suggestion
The GitHub token will expire Jan 7, 2026, at which time it will need to be replaced. This token grants this EC2 instance only the ability to pull down the Prolegis/congress repo. Eventually, we should add a GitHub action so that when Prolegis/congress is updated, the changes get deployed to this EC2 server. 
```

But wouldn't it expire sooner than described since this was set up months ago?

> Eventually, we should add a GitHub action so that when Prolegis/congress is updated, the changes get deployed to this EC2 server. 

If this is important then it should be ticketed, not stated as a side-note in a readme, else it'll never get done. If it isn't important or isn't planned to be done, I'd prefer to just not mention it here at all. I'm also not sure we want to automate pulling upstream changes as that would risk breaking our integration.

Author: Addressed here https://github.com/Prolegis/congress/pull/6/commits/6417f8e9ce2a3859f708061668492f35c3b0c87a#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L44

> But wouldn't it expire sooner than described since this was set up months ago?
I created a new token as I was updating this README. So the one we are using now should be good 👍

Reviewer: Do we really want to automatically pull upstream changes made to the parent repo? It seems like something that would potentially break our integration. It's like if we allowed Dependabot to just freely merge in every dependency version bump, even major ones, which would cause all sorts of issues.

Author: I think that's a good point actually. For now, let's make a note in the README that states that the user needs to automatically pull down the repo whenever it is changed, and that we should make this process be in a deploy script.

Author: I can open up a new PR to update the parts where `generate_bill_status_codes.sh` are using the GitHub token. So we can get this change in the README over the finish line. And then I will deactivate the GitHub token and remove it from 1Password.

> The credentials for the API keys can be found [here](https://my.1password.com/home#/everything/Search/xftvyzu5sqoqu3z67d5b3qewlih4v3tjt4h4qf5g72hxc47dkeey?itemListId=Congress+Repo+API+Keys). The credentials for the GitHub token can be found [here](https://my.1password.com/home#/everything/Search/p7c45mgv6bof7c5pagpgc3nolayn24v4dhlgovn7ysf3qqpcin3a?itemListId=token).

Reviewer: These links just take me to the 1pass homepage, but again, I don't think we should be linking directly to anything in 1pass in the first place

Author: Addressed here

https://github.com/Prolegis/congress/pull/6/commits/6417f8e9ce2a3859f708061668492f35c3b0c87a#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L44

This repository was originally developed by [GovTrack.us](https://www.govtrack.us) and the Sunlight Foundation in 2013 (see [Eric's blog post](https://sunlightfoundation.com/blog/2013/08/20/a-modern-approach-to-open-data/)) and is currently maintained by GovTrack.us and other contributors. For more information about data in Congress, see the [Congressional Data Coalition](https://congressionaldata.org/).

## EC2 Configuration
The purpose of this repository is to provide tools for generating congressional data that can be integrated into our Rails application. Since these tools are designed for local machine use, we cannot simply host the repository as an API to make requests.

Instead, we host the entire repository on an EC2 instance, enabling the following functionalities:
* Generating congressional data using the tools in this repository.
* Uploading the generated data to S3.
* Triggering requests to import this data into the Rails application (e.g., having the Rails app asynchronously read from S3 after hitting a specific endpoint).

### AWS Configuration
To enable S3 uploads, configure AWS access by running the following command in the terminal: `aws configure`

> You will be prompted to enter:

Reviewer 2: do we have the `.pem` file stored in 1Password? In case if anyone wants to ssh into the EC2 box?

Author: Good flag! I just added it now to 1Password. And I updated the README to indicate how to do so. https://github.com/Prolegis/congress/pull/6/commits/77b3656112aef841028ce70feedbe7e4473f28ba

## EC2 Configuration
The purpose of this repository is to provide tools for generating congressional data that can be integrated into our Rails application. Since these tools are designed for local machine use, we cannot simply host the repository as an API to make requests.

Instead, we host the entire repository on an EC2 instance, enabling the following functionalities:
* Generating congressional data using the tools in this repository.
* Uploading the generated data to S3.
* Triggering requests to import this data into the Rails application (e.g., having the Rails app asynchronously read from S3 after hitting a specific endpoint).

### AWS Configuration
To enable S3 uploads, configure AWS access by running the following command in the terminal: `aws configure`

You will be prompted to enter:
**1** **AWS Access Key ID**
**2** **AWS Secret Access Key**

⠀These credentials can be found in 1Password.
### Environment Variables
Set the following environment variables in the ~/.bashrc file to ensure proper functionality:

> * API_KEY_PRODUCTION - Enables interaction with the production Rails application.

Reviewer 2: are these the one under 
`Vote Import API Keys` ?? in 1Password

Author: @harish-prolegis no - that key is used for the GitHub action that is responsible for importing votes. Eventually we want to put all of that functionality in this EC2 instance but that is for a later day.

I wrote down what these files should be named in 1Password now. Hopefully that should help!

* API_KEY_STAGING - Enables interaction with the staging Rails application.
* API_KEY_DEMO - Enables interaction with the demo Rails application.

The credentials for the API keys can also be found in 1Password.

### Deploying changes
Right now, when a change is merged into the `master` branch, the repo is not deployed to the EC2 instance. Therefore, developers are required to go into the EC2 console and manually pull down the latest version of the congress repo. This is not ideal because it requires another extra step.

Eventually, we should add a GitHub action so that when Prolegis/congress is updated, the changes get deployed to this EC2 server.
A ticket to address this issue has been written here:  https://linear.app/prolegis-engineering/issue/PRO-3243/update-congress-repo-to-automatically-deploy-to-ec2-whenever-it-is

After editing `~/.bashrc`, apply the changes by running: `source ~/.bashrc`

### Accessing the console

> There are two ways to access the console for this EC2 instance. The first is to go through the AWS web portal itself. The second is to SSH into the instance. You can SSH into the instance using the .pem file in 1Password.

Reviewer 2: thank you. can you notate the name of the .pem file?

Author: Done
